Recommend `ssha256` cache hasher in FIPS docs (#87255)
elastic/elasticsearch: x-pack/docs/en/security/fips-140-compliance.asciidoc

[discrete]
=== Upgrade considerations

> [IMPORTANT] 

Author: Intellij keeps auto-formatting this. I'm assuming it's fine to drop them trailing spaces

you have external policies and tools that can ensure all user passwords for the reserved,
native, and file realms are longer than 14 bytes.

You must set the `xpack.security.authc.password_hashing.algorithm` setting to one of the
available `pbkdf_stretch_*` values.
When FIPS-140 mode is enabled, the default value for
`xpack.security.authc.password_hashing.algorithm` is `pbkdf2_stretch`.
See <<hashing-settings>>.

For cache hashing, salted `sha256` is recommended instead of `PBKDF2`. Though compliant
with FIPS-140-2, `PBKDF2` is -- by design -- slow, and thus not suitable as a cache hashing
algorithm. Cached credentials are never stored on disk, and salted `sha256` provides
an adequate level of security for in-memory credential hashing, without imposing prohibitive
performance overhead.

You should set the `cache.hash_algo` realm settings to `ssha256`. `ssha256` is the default value
for all `cache.hash_algo` realm settings.
See <<hashing-settings>>.

> You _may_ set the `cache.hash_algo` realm settings to any of the `pbkdf2` options, however you

Author: Not sure we need this paragraph, and happy to remove it if we don't think it adds value. Essentially, `pbkdf2` is a bad idea but it's not technically wrong and if a customer decides they want to use it and that the performance hit is acceptable, they can, from a FIPS standpoint. Wdyt?

Reviewer: I think this paragraph should go along with the previous one that says about `pbkdf2` not suitable for caching hashing.

<<security-api-change-password,change password>> APIs for the native and reserved realms.
Other types of realms are not affected and do not require any changes.

[discrete]

> ==== Cached Password Hashing

Author: I went with "Cached" instead of "In-Memory" because the section is specifically about cached passwords and cache settings, and we have no cases of in-memory password hashing that _isn't_ cache hashing.

> ==== Cached Password Hashing

Reviewer: I think we need add a dedicated anchor for this section in addition to the new separate title to be more resistant to potential future changes to page structure or wording. 
```suggestion
[[fips-cached-password-hashing]]
==== Cached Password Hashing
```

Author: will add `[[fips-stored-password-hashing]]` to the other section as well

> For cache hashing, salted `sha256` is recommended instead of `PBKDF2`. Though
> compliant with FIPS-140-2, `PBKDF2` is -- by design -- slow, and thus not

Reviewer: ```suggestion
`ssha256` (salted `sha256`) is recommended for cache hashing. Though `PBKDF2` is
compliant with FIPS-140-2, it is -- by design -- slow, and thus not
```

generally suitable as a cache hashing algorithm. Cached credentials are never
stored on disk, and salted `sha256` provides an adequate level of security for
in-memory credential hashing, without imposing prohibitive performance overhead.
You _may_ use `PBKDF2`, however you should carefully assess performance impact
first. Depending on your deployment, the overhead of `PBKDF2` could undo most of
the performance gain of using a cache.

> You should set the `cache.hash_algo` realm settings to `ssha256`. `ssha256` is

Reviewer: ```suggestion
You should set the `cache.hash_algo` settings to `ssha256` or just leave them undefined since `ssha256` is
```

> the default value for all `cache.hash_algo` realm settings. See

Reviewer: I am aware this reuses the existing wording, but `cache.hash_algo` is not only just realm settings, both service account and API key are not realms. So maybe just remove `realm`.

<<security-api-change-password,change password>> APIs for the native and reserved realms.
Other types of realms are not affected and do not require any changes.

[discrete]
==== Cached Password Hashing
[[fips-cached-password-hashing]]

`ssha256` (salted `sha256`) is recommended for cache hashing. Though
`PBKDF2` is compliant with FIPS-140-2, it is -- by design -- slow, and
thus not generally suitable as a cache hashing algorithm. Cached
credentials are never stored on disk, and salted `sha256` provides an
adequate level of security for in-memory credential hashing, without
imposing prohibitive performance overhead. You _may_ use `PBKDF2`,
however you should carefully assess performance impact first. Depending
on your deployment, the overhead of `PBKDF2` could undo most of the
performance gain of using a cache.

> You should set all `cache.hash_algo` settings to `ssha256` or leave

Reviewer 2: ```suggestion
Either set all `cache.hash_algo` settings to `ssha256` or leave
```
